Add system role support

dain opened the pull request:
Add support for system level roles in addition to the existing catalog roles.  This updates grammar, adds active system roles to identity, and updates file based security to support roles.  Currently roles can only be provided via an extension point in the coordinator, and an SPI will be added in a follow up PR.

### Before
* ROLEs only exist within a catalog
* Catalog only support catalog specific roles
* By default a user runs privileges from ALL roles in all catalogs
* SET ROLE can be used to reduce the active roles in a catalog

### After
* ROLEs can be system (global) or within a catalog
* Catalogs can either support system or catalog specific roles, but not both at the same time 
* By default a user runs privileges from ALL system roles and ALL roles in all catalogs
* SET ROLE can be used to reduce the active system roles or active roles in a catalog
* Roles can be used in system and catalog level file based access control

### Most Significant Change
* Require that catalog be explicitly declared in `CREATE ROLE`, `DROP ROLE` and `SET ROLE` with  optional `IN my_catalog` clause.

djsstarburst commented on core/trino-parser/src/main/java/io/trino/sql/SqlFormatter.java:
```diff
@@ -1582,6 +1590,10 @@ protected Void visitGrantRoles(GrantRoles node, Integer context)
             if (node.getGrantor().isPresent()) {
                 builder.append(" GRANTED BY ").append(formatGrantor(node.getGrantor().get()));
             }
+            if (node.getCatalog().isPresent()) {
```
Helper method `void appendInCatalogIfExists(builder, node)`?

djsstarburst commented on core/trino-main/src/main/java/io/trino/security/AccessControlManager.java:
```diff
@@ -849,112 +849,119 @@ public void checkCanSelectFromColumns(SecurityContext securityContext, Qualified
-        checkCanAccessCatalog(securityContext, catalogName);
-
-        catalogAuthorizationCheck(catalogName, securityContext, (control, context) -> control.checkCanCreateRole(context, role, grantor));
+        if (catalogName.isPresent()) {
```
Helper method?

djsstarburst commented on core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java:
```diff
@@ -1753,11 +1756,12 @@ public void dropRole(Session session, String role, Optional<String> catalog)
-
-            metadata.dropRole(session.toConnectorSession(catalogName), role);
-            return;
+            if (catalogMetadata.getRoleSupport() == RoleSupport.CONNECTOR) {
```
Helper method taking a lambda as the last arg?

djsstarburst reviewed (approved):
This looks great, Dain!  I particularly admire how the commits are largely structured as isolated, self-contained changes, arrayed in a logical sequence.

My only comments are that in a number of places, adding helper methods would cut down on code duplication.  I hate to repeat even a small code body more than twice, especially if it throws.

findepi commented:
@dain can you possibly extend PR description with some short summary of what it does?

findepi commented:
> @dain can you possibly extend PR description with some short summary of what it does?

thanks

findepi commented:
cc @trinodb/maintainers

kokosing reviewed (commented):
Can we extract first 6 commits (up to (including) `Add metadata role exists method`) to separate PR so we can merge it sooner?

dain commented on core/trino-main/src/main/java/io/trino/security/AccessControl.java:
```diff
@@ -395,63 +395,63 @@ default void checkCanSetViewAuthorization(SecurityContext context, QualifiedObje
      * @throws AccessDeniedException if not allowed
      */
-    void checkCanSetRole(SecurityContext context, String role, String catalogName);
+    void checkCanSetCatalogRole(SecurityContext context, String role, String catalogName);
```
For system roles, we don't do a check like this, because it would require the security system to know which role grants the user has.  Since role grants are kept in the metadata system this requires that there be a coupling between metadata and security.  Instead, when setting a role, we simply get the granted system roles from the metadata, and then we check if that list contains the specified role.  I expect this will be ok since the default role setting is "all roles", so the get roles method must be fast (so no perf concerns), and this keeps metadata separate from the security system.

dain replied:
> Notice that the problem already exists on catalog level.
Yep, but catalog roles have tons of problems, but I work hard in this PR to not touch catalog roles.
> Notice that for catalogs we could simply get the granted catalog roles from the metadata, and then we check if that list contains the specified role
Someone wanted to do that in another PR, but at that point they should take a long look at catalog roles overall.

dain commented on core/trino-main/src/main/java/io/trino/connector/ConnectorManager.java:
```diff
@@ -37,6 +37,7 @@
 import io.trino.spi.classloader.ThreadContextClassLoader;
```
When a connector says "I use global roles", the role methods are never called on the connector.  The connector security system can inspect the "enabled system roles" in the identity for access control checks it it wants.  This is what the file based connector access control in this PR does.

kokosing replied:
> When a connector says "I use global roles", the role methods are never called on the connector. 

When a connector says "I use system roles" then role methods are never called on the system, right?

It is incoherent with what we have now with everything related to access control. We were always doing full cascade of checks. In order to have an access, system would grant access and catalog would need to grant access. So goes to row filters and column masks we capture them from both access control levels.

I think user need to has an option. I he configures both levels then they expect to use both levels of access control. Like using file based system access control together with Hive Sql standard today. This change will break such use cases.

dain replied:
> > When a connector says "I use global roles", the role methods are never called on the connector.

> When a connector says "I use system roles" then role methods are never called on the system, right?

No.  What I wrote is want I meant.  When a connector does not have it's own roles, then roles are handles at the system level, so the catalog isn't consulted about things like set role.

> I think user need to has an option. I he configures both levels then they expect to use both levels of access control. Like using file based system access control together with Hive Sql standard today. This change will break such use cases.

I tried implementing that.  It was insanely complex, and it was really confusing to reason about.   This is the third (maybe forth) rewrite of this feature.

kokosing replied:
> When a connector says "I use system roles" then role methods are never called on the system, right?

🤦 I wanted to say: _then role methods are never called on the connector_.

> I tried implementing that. It was insanely complex, and it was really confusing to reason about.

It is one time effort. If we go with your approach we will need live with this for years, so if design is bad, misleading or complex to users, plugin developers or Trino contributors then I think we are going to pay a bigger price for this in a long run.

Why it was that complex? Maybe we can find another way to solve this? The fact that system and connector access controls are independent and evaluated in a cascade way is simple. Changing this with introducing a bridge that sometimes connector is using system roles breaks that design in my opinion.

Is it all about the syntax? That it may be ambiguous? If it is a case then let's solve it at syntax layer. Making syntax clear has an advantage that it will be clear to everyone: users, contributors and design.

For example if you are introducing system roles then maybe we could have, one of these:
 -  `SYSTEM GRANT role TO ...`, `SYSTEM GRANT SELECT ...`
 - `GRANT SYSTEM role ...`, `GRANT SYSTEM SELECT...`
 - `GRANT role ... IN SYSTEM`, `GRANT SELECT ... IN SYSTEM`

We could replace `SYSTEM` with `GLOBAL`. Any referred role would be a system role.

Alternatively, if we believe that system roles are first class citizen and connector roles are second class, then we could change syntax only for connector roles to always require `IN catalog` to be always specified.

dain commented on core/trino-spi/src/main/java/io/trino/spi/security/AccessDeniedException.java:
```diff
@@ -458,24 +458,24 @@ public static void denyRevokeTablePrivilege(String privilege, String tableName,
+    public static void denyShowRoles()
     {
-        throw new AccessDeniedException(format("Cannot show roles from catalog %s", catalogName));
+        throw new AccessDeniedException("Cannot show roles");
```
Not sure there is much we can do here.  This is how most deny exceptions work.  In ConnectorAccessControl the default methods all deny, and since it is an interface there are is no way to have access to the catalog name.

dain commented:
@kokosing here are the additional commits I added for your comments.  I also fixed some stuff in line and fixed tests in line.
* Update documentation for role management commands 
* Remove redundant "in this catalog" from ConnectorAccessControl docs
* Add flag to enable old catalog-only role commands
* Add checkRoleExists helper method
* Add Metadata getRequiredCatalogHandle

These fix additional stuff I found while applying the changes:
* Cleanup SetSchemaAuthorizationTask
* Cleanup CreateSchemaTask
* Automatically add admin role to file based metastore

kokosing commented:
Let's extract first 4 commits as separate PR so we can merge it. I don't think I will be able to swallow 22 commits at once and provide decent quality review. So I suggest to divide and conquer here.

kokosing commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java:
```diff
@@ -838,7 +838,12 @@ public synchronized void dropRole(String role)
     @Override
     public synchronized Set<String> listRoles()
     {
-        return ImmutableSet.copyOf(readFile("roles", getRolesFile(), rolesCodec).orElse(ImmutableList.of()));
```
Maybe using `Stream.concat` would produce cleaner code?

kokosing commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java:
```diff
@@ -838,7 +838,12 @@ public synchronized void dropRole(String role)
     @Override
```
Let's mention in the commit message that it was actually a bug because `listRoleGrants` was already returning admin role.

kokosing commented on core/trino-main/src/test/java/io/trino/execution/TestCreateSchemaTask.java:
```diff
@@ -36,6 +36,7 @@
 import java.util.concurrent.CopyOnWriteArrayList;
 
 import static io.airlift.concurrent.MoreFutures.getFutureValue;
+import static io.trino.execution.CreateSchemaTask.internalExecute;
```
I would avoid static import here. `internalExecute` name alone is very ambiguous.

kokosing commented on testing/trino-tests/src/test/java/io/trino/tests/TestServer.java:
```diff
@@ -248,7 +248,7 @@ public void testVersionOnError()
         // fails during parsing
         checkVersionOnError("SELECT query that fails parsing", "ParsingException: line 1:19: mismatched input 'fails'. Expecting");
         // fails during analysis
-        checkVersionOnError("SELECT foo FROM some_catalog.some_schema.no_such_table", "TrinoException: line 1:17: Catalog 'some_catalog' does not exist");
```
Why we lost an information about the line? It looks like a regression.

kokosing commented on core/trino-main/src/main/java/io/trino/metadata/MetadataUtil.java:
```diff
@@ -202,6 +203,20 @@ public static boolean tableExists(Metadata metadata, Session session, String tab
+        }
+    }
+
+    public static void checkRoleExists(Session session, Node node, Metadata metadata, TrinoPrincipal principal, String catalog)
```
nit: Please move this method one method up, so they are ordered by the usage from high level detail to low level.

kokosing commented on core/trino-main/src/main/java/io/trino/execution/CreateRoleTask.java:
```diff
@@ -21,6 +21,7 @@
 import io.trino.spi.security.TrinoPrincipal;
```
It would be nice to add some tests to `TestHiveRoles` to make sure that this new syntax is effective.

dain replied:
I added more tests

kokosing reviewed (commented):
Reviewed up to `Add support for global role management` (excluding).

kokosing commented on core/trino-main/src/main/java/io/trino/execution/CreateRoleTask.java:
```diff
@@ -19,24 +19,36 @@
 import io.trino.metadata.Metadata;
```
> Add flag to enable old catalog-only role commands

This commit should be squashed with, or go just after, the commit that introduces `IN <catalog>` syntax. We need to have tests that covers that legacy option still works as expected.

dain replied:
It is too much work to move at this point.

kokosing reviewed (commented):
Reviewed: `Add GlobalSecurityMetadata for global role management`

At this point I would like to stop, and resolve existing comments (and possibly merge approved first few commits) before moving forward.

kokosing commented on core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorAccessControl.java:
```diff
@@ -70,6 +70,19 @@
 {
+    enum RoleSupport
+    {
+        GLOBAL, CONNECTOR
```
SYSTEM? To match SystemAccessControl.

kokosing commented on core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorAccessControl.java:
```diff
@@ -70,6 +70,19 @@
+     */
+    default RoleSupport getRoleSupport()
+    {
+        return RoleSupport.CONNECTOR;
```
Would there be any connector that would ever return here `GLOBAL`?

kokosing commented on lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/security/FileBasedAccessControl.java:
```diff
@@ -118,7 +118,7 @@ public FileBasedAccessControl(String catalogName, FileBasedAccessControlConfig c
     public RoleSupport getRoleSupport()
     {
-        return RoleSupport.CONNECTOR;
+        return RoleSupport.GLOBAL;
```
Why not define roles for file based connector access control? Why do we use system roles here?

kokosing replied:
Connector roles are not only used with HMS. They are other access control that expressed as Trino connector access control that have roles.

dain commented on core/trino-main/src/main/java/io/trino/Session.java:
```diff
@@ -546,10 +546,10 @@ public SecurityContext toSecurityContext()
         private String source;
         private String catalog;
```
I'm not sure which class you are talking about, but someone else can do that in another PR

dain reviewed (commented):
I applied the actionable comments

kokosing commented on core/trino-main/src/main/java/io/trino/sql/rewrite/ShowQueriesRewrite.java:
```diff
@@ -297,7 +297,9 @@ protected Node visitShowRoles(ShowRoles node, Void context)
             if (node.isCurrent()) {
                 accessControl.checkCanShowCurrentRoles(session.toSecurityContext(), catalog);
-                List<Expression> rows = metadata.listEnabledRoles(session, catalog.orElseThrow()).stream()
+                Set<String> enabledRoles = catalog.map(c -> metadata.listEnabledRoles(session, c))
```
catalog is present as it is checked above you can simply ` metadata.listEnabledRoles(session, catalog.get))`

dain replied:
The check above should have been removed.  I think it was a rebase conflict

kokosing commented on core/trino-main/src/main/java/io/trino/sql/rewrite/ShowQueriesRewrite.java:
```diff
@@ -297,7 +297,9 @@ protected Node visitShowRoles(ShowRoles node, Void context)
                 accessControl.checkCanShowCurrentRoles(session.toSecurityContext(), catalog);
-                List<Expression> rows = metadata.listEnabledRoles(session, catalog.orElseThrow()).stream()
+                Set<String> enabledRoles = catalog.map(c -> metadata.listEnabledRoles(session, c))
+                        .orElseGet(() -> session.getIdentity().getEnabledRoles());
```
Using `session.getIdentity().getEnabledRoles()` looks incorrect as they are system roles, not catalogs roles which expect here.

dain replied:
The catalog check should have been removed.  If the catalog is present, you get the catalog roles, otherwise you get the system roles

kokosing reviewed (commented):
Please also:
- fail with nice error message a query `GRANT ROLE ... IN CATALOG ...` when given catalog is using `RoleSupport.GLOBAL`.
 - fail with the nice error message a query `GRANT SELECT ... TO ROLE ..` when given catalog is using `RoleSupport.GLOBAL`.

kokosing commented:
How about using RoleSupport.GLOBAL as default? Also I would consider using this for hive sql-standard too. That would make role behavior coherent across the catalogs. Then we could even consider removing RoleSupport at all.

dain commented on testing/trino-testing/src/main/java/io/trino/testing/TestingSessionContext.java:
```diff
@@ -29,13 +32,27 @@ private TestingSessionContext() {}
+            selectedRole = new SelectedRole(Type.ROLE, Optional.of(enabledRoles.iterator().next()));
+        }
+        else {
+            selectedRole = new SelectedRole(Type.ALL, Optional.empty());
```
This is a test utility to quickly convert from a Session to a SessionContext, which never happens in real code.  The problem is you can't go back from these three enum value to a list of roles, so this code assumes that if you have only one role, then we select that one, and if you have multiple roles, then that represents ALL roles you could possibly have.  This works because trino only supports no roles at all, exactly one role, or all roles you have access to.  Of course this code could convert ALL to NONE or one role, if you have no  roles or just one role, but it doesn't matter for tests.

dain commented on core/trino-main/src/main/java/io/trino/execution/SetRoleTask.java:
```diff
@@ -58,20 +57,24 @@ public String getName()
         }
         SelectedRole.Type type = toSelectedRoleType(statement.getType());
-        stateMachine.addSetRole(catalog.get(), new SelectedRole(type, statement.getRole().map(c -> c.getValue().toLowerCase(ENGLISH))));
+        stateMachine.addSetRole(catalog.orElse("system"), new SelectedRole(type, statement.getRole().map(c -> c.getValue().toLowerCase(ENGLISH))));
```
One of the core ideas of this design is we don't change the client protocol, because that takes forever to propagate through all of the client implementations.  Instead, we signal system roles using the system catalog, which is owned by Trino.  For existing clients that support roles, like the JDBC driver, they can simply set roles with `system:my_role`.

As for the QueryStateMachine role fields, I think it is clearer to use the existing method.  If we seperate out this one value of the role map, then we have to add cross checks to the methods which is just a pain, and error prone, without much value.

dain commented on core/trino-main/src/main/java/io/trino/server/HttpRequestSessionContextFactory.java:
```diff
@@ -207,11 +214,33 @@ public Identity extractAuthorizedIdentity(
+        if (selectedRole.getType() == Type.NONE) {
+            return identity;
+        }
+        Set<String> enabledRoles = metadata.listEnabledRoles(identity);
```
Yes.  This call must be fast or cached.  We load all of the enabled roles during session creation to ensure we get a consistent snapshot of roles for the query.

dain commented on core/trino-main/src/main/java/io/trino/server/HttpRequestSessionContextFactory.java:
```diff
@@ -243,22 +272,39 @@ private Identity buildSessionIdentity(Optional<Identity> authenticatedIdentity,
+    private static Optional<SelectedRole> parseSystemRoleHeaders(ProtocolHeaders protocolHeaders, MultivaluedMap<String, String> headers)
+    {
+        return parseProperty(headers, protocolHeaders.requestRole()).entrySet().stream()
+                .filter(entry -> entry.getKey().equalsIgnoreCase("system"))
```
See above.  We didn't want to change the protocol and this does clearly communicate the intent

dain commented:
Superseded by #9087
